processor/src/test/java/org/mapstruct/ap/test/bugs/_1005/Issue1005Test.java

```diff
@@ -0,0 +1,103 @@
+    Order.class,
+    OrderDto.class
+})
+public class Issue1005Test {
+
+    @WithClasses(Issue1005ErroneousAbstractResultTypeMapper.class)
+    @Test
+    @ExpectedCompilationOutcome(value = CompilationResult.FAILED,
+        diagnostics = {
+            @Diagnostic(type = Issue1005ErroneousAbstractResultTypeMapper.class,
+                kind = javax.tools.Diagnostic.Kind.ERROR,
+                line = 30,
+                messageRegExp = ".*\\.AbstractEntity is an abstract class that cannot be used to create an instance " +
+                    "for: .*\\.HasKey.")
+        })
+    public void shouldFailDueToAbstractResultType() throws Exception {
+    }
+
+    @WithClasses(Issue1005ErroneousAbstractReturnTypeMapper.class)
+    @Test
+    @ExpectedCompilationOutcome(value = CompilationResult.FAILED,
+        diagnostics = {
+            @Diagnostic(type = Issue1005ErroneousAbstractReturnTypeMapper.class,
+                kind = javax.tools.Diagnostic.Kind.ERROR,
+                line = 29,
+                messageRegExp = "The return type .*\\.AbstractEntity is abstract. Provide non abstract result type or" +
+                    " a factory method.")
+        })
+    public void shouldFailDueToAbstractReturnType() throws Exception {
+    }
+
+    @WithClasses(Issue1005ErroneousInterfaceResultTypeMapper.class)
+    @Test
+    @ExpectedCompilationOutcome(value = CompilationResult.FAILED,
+        diagnostics = {
+            @Diagnostic(type = Issue1005ErroneousInterfaceResultTypeMapper.class,
+                kind = javax.tools.Diagnostic.Kind.ERROR,
+                line = 30,
+                messageRegExp = ".*\\.HasPrimaryKey is an interface and cannot be used to create an instance for: " +
+                    ".*\\.HasKey."),
+            @Diagnostic(type = Issue1005ErroneousInterfaceResultTypeMapper.class,
+                kind = javax.tools.Diagnostic.Kind.ERROR,
+                line = 31,
+                messageRegExp = "No implementation type is registered for return type .*\\.HasKey.")
+        })
+    public void shouldFailDueToInterfaceResultType() throws Exception {
+    }
+
+    @WithClasses(Issue1005ErroneousInterfaceReturnTypeMapper.class)
+    @Test
+    @ExpectedCompilationOutcome(value = CompilationResult.FAILED,
+        diagnostics = {
+            @Diagnostic(type = Issue1005ErroneousInterfaceReturnTypeMapper.class,
+                kind = javax.tools.Diagnostic.Kind.ERROR,
+                line = 29,
+                messageRegExp = "The return type .*\\.HasKey is interface. Provide non abstract/interface result type" +
+                    " or a factory method."),
+            @Diagnostic(type = Issue1005ErroneousInterfaceReturnTypeMapper.class,
+                kind = javax.tools.Diagnostic.Kind.ERROR,
+                line = 29,
+                messageRegExp = "No implementation type is registered for return type .*\\.HasKey.")
+        })
+    public void shouldFailDueToInterfaceReturnType() throws Exception {
```

Comment:
Should we really report two very similar error messages for the same code line? I think the second message isn't realy required here, as the one above has more context information and hints about what to do. Plus, some compilers (e.g. JDK) swallow all but the first error message for the same location. See `JdkCompilingStatement.filterExpectedDiagnostics`... :)